pkg/controller/queuejob/queuejob_controller_ex.go

```diff
@@ -1105,342 +1107,354 @@ func (qjm *XController) ScheduleNext() {
-			if qjm.quotaManager != nil && quotaFits {
-				qjm.quotaManager.Release(qj)
-			}
-			go qjm.backoff(qj, dispatchFailedReason, dispatchFailedMessage)
 		}
+		return nil
+	})
+	if apiErrors.IsNotFound(err) {
+		klog.Warningf("[ScheduleNext] app wrapper '%s/%s' not found skiping dispatch", qj.Namespace, qj.Name)
+		return
+	}
+	if err != nil {
+		klog.Warningf("[ScheduleNext] failed to dispatch the app wrapper '%s/%s', err= %v", qj.Namespace, qj.Name, err)
+		klog.Warningf("[ScheduleNext] retrying dispatch")
+		qjm.qjqueue.AddIfNotPresent(qj)
 	}
 }
 
 // Update AppWrappers in etcd
 // todo: This is a current workaround for duplicate message bug.
-func (cc *XController) updateEtcd(qj *arbv1.AppWrapper, at string) error {
-	//apiCacheAWJob, e := cc.queueJobLister.AppWrappers(qj.Namespace).Get(qj.Name)
-	//
-	//if (e != nil) {
-	//	klog.Errorf("[updateEtcd] Failed to update status of AppWrapper %s, namespace: %s at %s err=%v",
-	//		apiCacheAWJob.Name, apiCacheAWJob.Namespace, at, e)
-	//	return e
-	//}
-
-	//TODO: Remove next line
-	var apiCacheAWJob *arbv1.AppWrapper
-	//TODO: Remove next line
-	apiCacheAWJob = qj
-	apiCacheAWJob.Status.Sender = "before " + at // set Sender string to indicate code location
-	apiCacheAWJob.Status.Local = false           // for Informer FilterFunc to pickup
-	if _, err := cc.arbclients.ArbV1().AppWrappers(apiCacheAWJob.Namespace).Update(apiCacheAWJob); err != nil {
-		klog.Errorf("[updateEtcd] Failed to update status of AppWrapper %s, namespace: %s at %s err=%v",
-			apiCacheAWJob.Name, apiCacheAWJob.Namespace, at, err)
-		return err
-		//	} else {  // qjj should be the same as qj except with newer ResourceVersion
-		//		qj.ResourceVersion = qjj.ResourceVersion  // update new ResourceVersion from etcd
+func (cc *XController) updateEtcd(currentAppwrapper *arbv1.AppWrapper, caller string) (*arbv1.AppWrapper, error) {
+	klog.V(4).Infof("[updateEtcd] trying to update '%s/%s' called by '%s'", currentAppwrapper.Namespace, currentAppwrapper.Name, caller)
+	currentAppwrapper.Status.Sender = "before " + caller // set Sender string to indicate code location
+	currentAppwrapper.Status.Local = false               // for Informer FilterFunc to pickup
+	updatedAppwrapper, err := cc.arbclients.ArbV1().AppWrappers(currentAppwrapper.Namespace).Update(currentAppwrapper)
+	if err != nil {
+		return nil, err
 	}
-	klog.V(10).Infof("[updateEtcd] AppWrapperUpdate success %s at %s &qj=%p qj=%+v",
-		apiCacheAWJob.Name, at, apiCacheAWJob, apiCacheAWJob)
-	//qj.Status.Local  = true           // for Informer FilterFunc to ignore duplicate
-	//qj.Status.Sender = "after  "+ at  // set Sender string to indicate code location
-	return nil
+	if larger(updatedAppwrapper.ResourceVersion, currentAppwrapper.ResourceVersion) {
```

Comment:
we are using locks to work on single AW, curios to know why do we still need larger check?

Reply:
I think this block is superfluous. We'll remove.